Create parser for Jefferson County PDF from AL (#184)

adityasharad opened the pull request:
# Parse Jefferson County PDF for AL

| Key Details |
|-|
| Resolves #184 |
State: AL |
Site: Jefferson County |

## Notes
* The Jefferson County data is [a PDF](https://www.jcdh.org/SitePages/Misc/PdfViewer?AdminUploadId=1368). To parse it, this PR uses one of the strategies agreed on in the discussion on #184:
  * Use `pdftohtml`, invoked from Python, to convert the PDF into XML.
  * Use `lxml` to process the XML in Python into NDJSON.
* The document does not have a consistent structure for separating each vaccine provider and site, only header images for each provider (which are not parsed as text).
  * This PR uses line-spacing heuristics to group XML lines of text into vaccine providers
  and vaccine sites for each provider.
  * This is arguably the most complex part of the parsing logic, found in the function `_group_text_by_provider`. I have attempted to provide extensive inline comments and am happy to explain more if necessary. Switching the log level to `DEBUG` will produce verbose logs, which are useful to debug the grouping logic.
  * Sadly this means we don't have all the provider names, unless we run OCR on the header images or look the names up through some other means.
* For security, the XML parsing runs with `resolve_entities=False` and `huge_tree=False` to disable various forms of XML entity expansion. This is not 100% protection against untrusted XML but is intended to mitigate potential DoS.

### Open questions for reviewers
- [x] Should the parser attempt to remove text lines that are not associated with a vaccine site (e.g. the `VACCINE FAQ` lines below), or is the normaliser the right place for this?
- [x] Should the parser attempt to extract address and phone number from the grouped details, or is the normaliser the right place for this?

## Data sample

This corresponds to page 1 of the [PDF](https://www.jcdh.org/SitePages/Misc/PdfViewer?AdminUploadId=1368). (Edited to show updated output, after addressing review comments.)
```json
{"page": 1, "provider": 7, "link": "http://www.jcdh.org/", "details": ["1400 Sixth Avenue South,", "Birmingham, AL 35233", "(205) 858-2221"]}
{"page": 1, "provider": 11, "link": "https://www.uabmedicinevaccine.org/", "details": ["Cathedral of the Cross AOH Church", "1480 Center Point Parkway,", "Birmingham, AL 35215"]}
{"page": 1, "provider": 11, "link": "https://www.uabmedicinevaccine.org/", "details": ["Parker High School", "400 Reverend Abraham Woods Jr. Blvd,", "Birmingham, AL 35204"]}
{"page": 1, "provider": 11, "link": "https://www.uabmedicinevaccine.org/", "details": ["Hoover Metropolitan Complex", "100 Ben Chapman Drive,", "Hoover, AL 35244"]}
{"page": 1, "provider": 11, "link": "https://www.uabmedicinevaccine.org/", "details": ["UAB Hospital Highlands", "1130 12th Street South,", "Birmingham, AL 35205"]}
{"page": 1, "provider": 12, "link": "https://www.cahabamedicalcare.com/covid19", "details": ["Cahaba Medical Care Fairfield", "6900 Grasselli Road,", "Fairfield, AL 35064", "(205) 791-5048"]}
{"page": 1, "provider": 12, "link": "https://www.cahabamedicalcare.com/covid19", "details": ["Cahaba Medical Care Bessemer", "975 9th Ave SW, Suite 310,", "Bessemer, AL 35022", "(205) 277-2358"]}
{"page": 1, "provider": 12, "link": "https://www.cahabamedicalcare.com/covid19", "details": ["Cahaba Medical Care Ensley", "1925 Avenue E Ensley,", "Birmingham, AL 35218", "(205)788-5164"]}
{"page": 1, "provider": 13, "link": "https://birminghamairport.timetap.com/#/", "details": ["5900 Messer Airport Hwy,", "Birmingham, AL 35212"]}
{"page": 1, "provider": 14, "link": "https://medspluscovidvaccineclinics.timetap.com/#/", "details": ["2337 Bessemer Rd,", "Birmingham, AL 35208"]}
{"page": 1, "provider": 15, "link": "https://myturn.timetap.com/#/", "details": ["4500 Katies Way,", "Bessemer, AL 35022", "(205) 858-2221"]}
```

## Before Opening a PR
- [x] I tested this using the CLI (e.g., `poetry run vaccine-feed-ingest <state>/<site>`)
- [x] I ran auto-formatting: `poetry run tox -e lint-fix`

bryanculbertson commented on .github/workflows/lint.yml:
```diff
@@ -33,7 +33,7 @@ jobs:
         run: |
           sudo apt-get update
-          sudo apt-get install libbz2-dev liblzma-dev libreadline-dev libsqlite3-dev
+          sudo apt-get install libbz2-dev liblzma-dev libreadline-dev libsqlite3-dev pdftohtml
```
Can you also add this to the wiki: https://github.com/CAVaccineInventory/vaccine-feed-ingest/wiki/Manual-Development-environment-setup

And to the setup.sh script: https://github.com/CAVaccineInventory/vaccine-feed-ingest/blob/main/setup.sh

adityasharad replied:
Added to `setup.sh`. If you're willing to temporarily give me write access to the wiki, I'll add it there too.

bryanculbertson replied:
I don't have permission to adjust permissions, but I will add it to the wiki. Thanks!

bryanculbertson reviewed (approved):
This is a complex parse, thank you for tackling it. I suspect there will be a a couple more iteration on this one as you work on the normalizer.

As for your questions, I think you should to a low pass filter in the parser to remove rows that are not related to locations. I think this will make the normalizer easier to write.

I also think you should leave the address and phone number parsing to the normalizer, unless having knowledge of the PDF or XML structure would help with that.

bryanculbertson reviewed (changes requested):
Code is approved, but waiting to merge until you can update the setup instructions, and get the tests passing

adityasharad commented:
Thanks @bryanculbertson for the quick review! I believe I have addressed all comments, and will await a fresh CI run to verify.
Leaving address and phone parsing to the normaliser sounds good to me, happy to take that up next.

adityasharad commented:
Super Linter errors were already present in `setup.sh`, but presumably surfaced because I modified the file. Now fixed in latest commit.

adityasharad commented:
Test failures are in `sc/arcgis/normalize.py`, and I think unrelated. Shall I attempt to fix those and push (will need reapproval for Actions to run), or are you happy to merge as-is?

eliblock commented:
Hi @adityasharad - I'm going to override CI on this one.

The `Test & Lint / Run all-stages for modified runners (pull_request)` is intended to run all runners which were modified in your PR, but it looks like we're being a bit over-eager here.
